### ISSUE TITLE
onMouseEnter/onMouseLeave events for overlapping Touchables don't dispatch correctly (Regression from 0.60)

### ISSUE THREAD

**Issue by KAnder425** [microsoft/react-native-windows#5901]
## Environment
1. `npx react-native --version`:
4.10.0
2. `npx react-native run-windows --info`:
  System:
    OS: Windows 10 10.0.19041
    CPU: (8) x64 Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
    Memory: 4.15 GB / 15.92 GB
  Binaries:
    Node: 12.18.3 - C:\Program Files\nodejs\node.EXE
    Yarn: 1.16.0 - C:\Program Files (x86)\Yarn\bin\yarn.CMD
    npm: 6.14.6 - C:\Program Files\nodejs\npm.CMD
  npmPackages:
    react-native: 0.62.2 => 0.62.2
    react-native-windows: 0.62.5 => 0.62.5
  Installed UWP SDKs:
    10.0.14393.0
    10.0.17134.0
    10.0.17763.0
    10.0.18362.0
3. `reg query "HKLM\SOFTWARE\Microsoft\Windows\CurrentVersion\AppModelUnlock"`
HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\AppModelUnlock
    AllowDevelopmentWithoutDevLicense    REG_DWORD    0x1
    AllowAllTrustedApps    REG_DWORD    0x1

- Target Platform Version(s): 10.0.18362.0
- Target Device(s): UWP
- Visual Studio Version: Visual Studio 2019

## Steps To Reproduce
1. Run the mouse sample in Playground 
2. Hover over the section of the orange overlay View that overlaps the blue TouchableHighlight

The onMouseLeave event is fired for the TouchableHighlight even though the mouse has not left the underlying TouchableHighlight rect.

## Expected Results
The same behavior as in 0.60.

Our app has several instances of this scenario, with the worst impact being a continuous flicker of hover treatment back and forth. 

## Screenshots (gifs):
Behavior in 0.62 (and latest master):

![mouse touchables 0 62](https://user-images.githubusercontent.com/5695388/91893508-165a3600-ec49-11ea-918c-b703431a792b.gif)

Behavior in 0.60 (latest):

![mouse touchables 0 60](https://user-images.githubusercontent.com/5695388/91893521-19edbd00-ec49-11ea-99c3-d4a236fd1111.gif)

**Comment by NickGerleman** [comment]
@KAnder425 I believe that behavior was changed more intentionally after a few folks thought it was a bug. See https://github.com/microsoft/react-native-windows/issues/4274

It's useful to know their was reliance on previous behavior. We've been trying to make some behavior changes more controllable (E.g. with quirk settings) to avoid scenarios like this, where something seen as a bugfix breaks someone. It looks like we missed this. I'm not sure if a change allowing previous behavior in this instance could be brought back safely.

Mouse events and keyboard events are kind of in an unfortunate place right now. They were implemented, but often with loose/undocumented contracts, or behavior that conflicts with other platforms. Moving forward there has been a lot of recent effort to drive formalizing what keyboard/mouse APIs and contracts should look like across Windows, MacOS, iPadOS, etc.

**Comment by KAnder425** [comment]
Yes. I just saw 4274 :-)

**Comment by KAnder425** [comment]
Ok, I understand that this is a design decision, follow the virtual DOM, or follow the actual XAML.  I prefer following the XAML that's being displayed on the screen, but to be honest, I don't know which is the best approach because it depends on the mix of apps being created in RNW.  Either way, we need to pick one design choice and then widely communicate the choice.

In the mouse.tsx sample scenario, the current behavior looks broken.  

I will investigate the difficulty of recoding our overlapping views in our app to match the DOM, but I'm guessing that at some point someone will come up with complex enough scenarios that this won't be possible.

**Comment by chrisglein** [comment]
@kikisaints as part of your mouse reconciliation can you advice on what behavior we want here and how we want to communicate it?

**Comment by kikisaints** [comment]
Is there a reason we cannot get the same behavior in RNW that we do in vanilla native WinUI apps?
This is what I'm referring to for reference:
![image](https://im2.ezgif.com/tmp/ezgif-2-63e137502c2a.gif)

If not, I would say that if the 0.60 behavior is what's expected by most React Native devs than we should be doing that.

**Comment by aschultz** [comment]
@KAnder425 I'm not sure what you mean by "follow the virtual DOM, or follow the actual XAML". The hierarchy of elements is the same, be it in JS or XAML. The problem is that the 0.60 algorithm ignores that hierarchy completely and treats the mouse as being "inside" every element under the cursor position at any depth.

The fix I implemented changes it to match [web](https://w3c.github.io/uievents/#mouseenter), where the mouse is limited to being "inside" a single subtree of elements.

@kikisaints visual denotes an even more limited situation, where the mouse is limited to being "inside" only the top-most element.

I definitely think the 0.60 behavior is the wrong direction. As called out in #4274 it makes layering interactive elements impossible without introducing poor UX.

@kikisaints top-most-only is also problematic, because there are a number of cases where the element for which we want to capture enter/leave is *not* the top-most. Consider a simple button with text in it. I want to highlight the button when the mouse is over it. When the mouse enters the button, it should get a mouse enter. If we then hover the text inside, the text should get a mouse enter, but the button should not get a mouse leave.

The web behavior provides solutions to both of these problems. It also matches web, so it reduces the difficulty in porting code between the platforms and should be more familiar to React developers.

**Comment by KAnder425** [comment]
(I'm sorry that I missed this thread over the holidays.)
When I said "follow the virtual DOM or follow the actual XAML", I was referring to the difference in behavior between the 2 algorithms with respect to overlapping elements. The web approach will only consider overlapping elements within the RN subtree of the topmost element, while the v0.60 approach considered all of the overlapping UI elements regardless of location in the RN tree (as in Windows).
 
It's a design choice and either approach is workable as long as it is communicated clearly and doesn't change. 
On the other hand, either choice introduces its own complications. 
 
The problem that we encountered with the switch to the web approach came with a complex component that had sibling elements that were styled to render visually overlapped. This broke the hover behavior in the following way…
 
[User Story 4414310: Focus flickers when hovering over a date in the mini calendar (Left Nav Pane) - Boards (visualstudio.com)](https://office.visualstudio.com/Universal%20Outlook/_workitems/edit/4414310)
 
In the gif below, each 'week' (MiniCalendarWeek) in the the 'mini-Calendar' control on the left is implemented as a TouchableHighlight that is positioned (absolute) underneath the 7 'days' (MiniCalendarItem) which are also TouchableHighlights.
![flicker on hover 2](https://user-images.githubusercontent.com/5695388/104080712-93408200-51de-11eb-882e-a937657e99fa.gif)

eg..  
```
<View>
   <MiniCalendarWeek/> 
   …and then 7
  <MiniCalendarItem/>
</View>
```
 
It wasn't possible to rearrange the RN tree to fit the 'web approach' to mouse event handling in v0.62+ due to the level of reuse of the underlying MiniCalendarItem in other views and some Flex layout restrictions. However, we were able to work around this with some rather complex code to track the hover state up at a higher level in that component.
 
I realize that this is a long explanation, and even after all this I do agree that for RN/RNW, the web approach is probably the right direction, but I think this illustrates the challenges in dealing with mouse/pointer tracking in a x-plat framework. 
 
BTW, I'd agree that the 'top-most-only' approach as described above is insufficient.

**Comment by aschultz** [comment]
@KAnder425 Thanks for the explanation!

I can see how the change would break you. My solution would have been to re-arrange as you suggested -- make the days children of the week, so that when they are hovered the mouse is "inside" both the day and the parent week. If you needed a child to expand out to bigger than the row without displacing anything (say, hovering a day adds a bigger circle), you could probably use a transform or negative margins on the child to overflow.

For comparison, see #4247 for my use case (Xbox app). We have a full-window scroll view that contains the majority of the app content. We then have a nav bar that's a sibling to the scrollview, positioned at the top of the page but visually overlapping the scroll view. It has acrylic applied so underlying content bleeds through a little bit. 

The problem we have with the 0.60 algorithm is that hovering buttons in the nav bar will also hover content in the scroll view. Same for our notification tray, which is a sibling to both that pops out to overlay the main page content -- hover stuff in the tray, it also hovers the main page content. Not what we want.

It happens because despite being separate subtrees, they are both under the mouse cursor.

That's the crux of it. The algorithm is indiscriminate and can't be fixed with re-arranging. Every single thing under the cursor is "hovered", regardless of the DOM hierarchy, regardless of z-index.

How would you even start to work around that? It's a global problem. Do you add tracking code to every single hoverable element in the entire tree just so it could filter out spurious events?


That's why I think the web algorithm is the only realistic choice here. "Mouse is in everything under the cursor" (0.60) wholly prevents segregating content into layers. "Mouse is in topmost element under the cursor" makes it difficult to have nested effects within a layer. "Mouse is in topmost element under cursor and its ascendants" gives the flexibility to choose based on how you arrange your content. If you want something to be layered as "top-most" and capture events, you make it a sibling and position it on top. If you want events to fire on multiple things under the cursor, nest them as parent-child. And you can mix and match those throughout your app.

FWIW, "Mouse is in topmost under cursor and its ascendants" is what UWP does for PointerEntered/PointerExited too.

**Comment by kikisaints** [comment]
@aschultz just to help clarify, when you say "topmost under the cursor and it's ascendants" how would you expect that to actually function? You mention that it should be what UWP does, which is what I posted in the short gif I provided (that's the native, non-RNW behavior in UWP apps today), but you mentioned that behavior isn't quite right to solve the issue. Can you elaborate more on what that solution would look like on a more granular level?

Conversely, if what we did in 0.60 is exactly what we need, I can call this a reversion and tag it as such in triage.

**Comment by aschultz** [comment]
@kikisaints Regarding your gif -- I don't know how you put together the code for it, but it's not the behavior I observe on UWP. I see UWP acting the same as the web algorithm -- the pointer is "in" the top-most element under the cursor and all of its ascendants in the tree.

Here's a vanilla C# app to prove my point: https://github.com/aschultz/hittest

The structure of the XAML for that sample is very simple and aligns with the HTML I posted earlier:
```
<Grid x:Name="RootGrid">
    <Grid x:Name="A">
        <Grid x:Name="AA"><TextBlock>AA</TextBlock></Grid>
        <TextBlock>A</TextBlock>
    <Grid>
    <Grid x:Name="B">
        <Grid x:Name="BB"><TextBlock>BB</TextBlock></Grid>
        <TextBlock>B</TextBlock>
    <Grid>
    <TextBlock>Root</TextBlock>
</Grid>
```
Each Grid has PointerEntered/PointerExited handlers attached that simply decrease/increase the background color when fired.

Here's a gif of the mouse moving across the elements:
![hittest](https://user-images.githubusercontent.com/1130900/105805192-5999a600-5f56-11eb-8eaa-ad14281ae34d.gif)

As the mouse moves from `Root` to `A` to `AA`, UWP fires three PointerEntered events and no PointerExited. Thus when the pointer is over `AA` it is "inside" `AA` but also still inside its ascendants `A` and `Root`:
```
PointerEntered - Sender: RootGrid             - Source: RootGrid            
PointerEntered - Sender: A                    - Source: A                   
PointerEntered - Sender: AA                   - Source: AA       
```

When the mouse moves from `AA` to `B`, UWP fires PointerExited for `A` and `AA` and fires PointerEntered for `B`. Thus the mouse is no longer "inside" the `A` subtree; it is now inside the `B` subtree (and still inside `Root`):
```
PointerExited  - Sender: AA                   - Source: B                   
PointerExited  - Sender: A                    - Source: B                   
PointerEntered - Sender: B                    - Source: B  
```

The mouse is able to move around inside `B` even past the extents of the underlying `Root` and still be considered "inside" the `Root->B` subtree. Once the mouse move out of `B` to the page background, it has left both `B` and `Root` and the corresponding PointerExited events fire:
```
PointerExited  - Sender: B                    - Source:                     
PointerExited  - Sender: RootGrid             - Source:                     
```

**The UWP behavior described here is the same as what I implemented for RNW in 0.62, as you can see in the OP's gif.**

TLDR
---
Please do not revert to 0.60. The behavior already in 0.62+ is what we need. This bug should be closed as By Design. The OP has code that took a dependency on the broken behavior in 0.60 -- they should fix their code, not require RNW to remain broken.

**Comment by kikisaints** [comment]
I see, thank you @aschultz. The goal is to have what that UWP does, be the same across into RNW. So, if this is the case with the 0.62 update, then that's the correct behavior.

@KAnder425 I will be closing this issue as by design.
